libV2/server/API/metrics/listMetrics.js

```diff
@@ -0,0 +1,82 @@
+const errors = require('../../../errors');
+const { serviceToWarp10Label, operationToResponse } = require('../../../constants');
+const { convertTimestamp } = require('../../../utils');
+const { client: warp10 } = require('../../../warp10');
+
+const emptyOperationsResponse = Object.values(operationToResponse)
+    .reduce((prev, key) => {
+        prev[key] = 0;
+        return prev;
+    }, {});
+
+const metricResponseKeys = {
+    buckets: 'bucketName',
+    accounts: 'accountId',
+    users: 'userId',
+    service: 'serviceName',
+};
+
+async function listMetric(ctx, params) {
+    const labelName = serviceToWarp10Label[params.resource];
+    const resources = params.body[params.resource];
+    const [start, end] = params.body.timeRange
+        .map(convertTimestamp)
+        .map(v => v.toString());
+
+    // A separate request will be made to warp 10 per requested resource
+    const results = await Promise.all(
+        resources.map(async resource => {
+            const labels = {};
+            labels[labelName] = resource;
+            const options = {
+                params: {
+                    start,
+                    end,
+                    labels,
+                },
+                macro: 'utapi/getMetrics',
+            };
+            const res = await warp10.exec(options);
+            if (res.result.length === 0) {
+                ctx.logger.error('unable to retrieve metrics', { resource, type: params.resource });
+                throw errors.InternalError;
+            }
+            return {
+                resource,
+                metrics: JSON.parse(res.result[0]),
+            };
+        }),
+    );
+
+    // Convert the results from warp10 into the expected response format
+    const resp = results
+        .map(result => {
+            const operations = Object.entries(result.metrics.operations)
+                .reduce((prev, [key, value]) => {
+                    prev[operationToResponse[key]] = value;
+                    return prev;
+                }, {});
+
+            const metric = {
+                ...result.metrics,
+                timeRange: params.body.timeRange,
+                operations: {
+                    ...emptyOperationsResponse,
+                    ...operations,
+                },
+            };
+            metric[metricResponseKeys[params.resource]] = result.resource;
+            return metric;
+        });
+
+    // Unwrap the response if only one resource is passed
```

Comment:
Is there one returned resource if and only if only one resource is requested? If so, that looks fine as long as the contract with the API usage is clear, otherwise better keep a consistent API and always return the same data type (array of results, one item per resource?)

By the way, if the API takes an array of resources to query and there's only one resource in that array, I'd rather return an array of one result than a single unwrapped result, also for consistency.

Not sure if this all makes sense, feel free to ignore :smile:

Reply:
I agree that a consistent return type is better, but we need to maintain API compatibility with v1.

Reply:
OK, then let's keep as is.